segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerKeyIndex.java

```diff
@@ -209,6 +209,26 @@ public void close() {
         }
     }
 
+    /**
+     * Finds the Bucket Offset for the given Key Hash directly from the index (excluding the cache). If the index contains
+     * an updated offset for this Bucket (i.e., as a result of a compaction), the cache will be automatically updated in
+     * the process. If the index contains an obsolete offset for this Bucket (compared to the cache), the cache value
```

Comment:
Consider this scenario, for Key K:
1. We fetch K, and it was found at offset O1. Hence the cache contains K->O1.
2. We compact and move K to O2, then we truncate at offset T (O1<T<=O2). At this point, the index contains K->O2, but the cache still has K->O1.
3. We try to fetch K, but when reading the key data at offset O1, we get a `SegmentTruncatedException`. We need to get the updated position from the index (O2) and update the cache (if needed) and return the result.

Reply:
Your explanation makes sense, but I'm stuck at this:

> If the index contains an obsolete offset for this Bucket (compared to the cache), the cache value will be returned.

You scenario argues about the cache containing a stale value, but this sentence talks about the cache having fresher data compared to the index.

Reply:
Yes, there are two cases:
1. Cache contains outdated information. This happens when a compaction moved an entry from the back to the front, and nothing else happened to the key (no new updates in the meantime). In this case, the correct value is loaded from the index and stored in the cache, prior to returning it.
2. Cache contains the correct value. This can happen when there has been no compaction (so index value matches cache value) or there has been a recent update to this key that is not yet reflected in the index. In this case, the value from the index is discarded and the cached value returned.

Please note that it is impossible for us to determine with exact accuracy if a key in the cache has become obsolete. If you look above (237-255) you'll see that we resort to this if one of the following happens:
- We encounter a truncated exception (we know for sure this is outdated) 
- We get a `null` entry from `bucketReader.find`. This is returned either because the entry has been deleted or if we have a multi-entry bucket (due to collisions) which had to be rearranged due to compaction (example below).

Let me explain the bucket rearrangement.
- Suppose we have 3 keys in a bucket: K1-K3.
- The bucket points to the offset of K3, and we use backpointers to link the rest: K3->K2, K2->K1.
- We compact and move K1 to the end, so the indexer will rearrange the bucket as K1->K3; K3->K2.
- The cache still points to K3, so if we look for K1 now (after reindexing), we will start at K3 and only find K3 and K2, but not K1. In this case we need to go directly to the index to get the new location of the bucket (K1) so that we may return the right result.

An alternative to this cache invalidation on demand would be to have a notification system from the indexer into the cache. But that would be unnecessarily complex, chatty and would introduce tight coupling between the components, making them much harder to test. I think that a small performance penalty (due to an unneeded retry) in a corner case is acceptable to this alternative.